framework/doc/content/syntax/UserObjects/index.md

```diff
@@ -37,6 +37,61 @@ and use the interface without the use of dynamic casts.
 - +DomainUserObject+: this object is capable of executing all the operations of
   a +ElementUserObject+, +InternalSideUserObject+, +SideUserObject+ and +InterfaceUserObject+.
 
+# Execution order
+
+Within an execution stage set by the [`execute_on`](SetupInterface.md) parameter, user objects are executed in
+the following order:
+
+1. `residualSetup` / `jacobianSetup`
+
+   If the current `execute_on` flag is either `EXEC_LINEAR` or `EXEC_NONLINEAR` the `residualSetup`
+   and `jacobianSetup` are called respectively in the following order
+
+   1. for objects derived from `ElementUserObject`,  `SideUserObject`, `InternalSideUserObject`
+      `InterfaceUserObject`, and  `DomainUserObject`.
+   2. for objects derived from `NodalUserObject`.
+   3. for objects derived from `ThreadedGeneralUserObject`.
+   4. for objects derived from `GeneralUserObject`.
+
+2. `initialize` is called for objects derived from `ElementUserObject`,  `SideUserObject`,
+   `InternalSideUserObject` `InterfaceUserObject`, and  `DomainUserObject` in that order.
+
+3. All active local elements are iterated over and objects derived from `ElementUserObject`,
+   `SideUserObject`, `InternalSideUserObject` `InterfaceUserObject`, and  `DomainUserObject` are
+   executed on elements, boundaries attached to the element, internal sides between elements, and on subdomain changes, in that order.
+   The order within each type group is determined through dependency resolution.
+
+4. `threadJoin` and `finalize` are called in the following order
+
+   1. for objects derived from `SideUserObject`
+   2. for objects derived from `InternalSideUserObject`
+   3. for objects derived from `InterfaceUserObject`
+   4. for objects derived from `ElementUserObject`
+   5. for objects derived from `DomainUserObject`
```

Comment:
So this inconsistent ordering is because of Griffin? Is the plan to merge this, use the execution order feature in Griffin where needed, and then fix this ordering to be consistent?

Reply:
Um, I don't have any plans to fix griffin. Why is this ordering inconsistent? I also suspect that after years of living with this ordering other apps rely on this ordering. We can certainly explore changing that order, but that'll be a different PR.

Reply:
Yeah I don't think we can get around this default and I'm fine with that

Reply:
If I recall correctly, we used to have `ElementUserObject` as a main object that gathers info from `SideUserObject`s in Griffin. I am not exactly sure what you mean inconsistency, maybe the ordering in setups? But I think after introducing `DomainUserObject`, we got rid of this dependency. Either way, I do not think it is wise to change something in this PR which is introducing a new feature for now.

Reply:
Just to clarify, my question is not a requirement for this PR - it should be merged regardless.

By "inconsistent" I meant that in all of the other steps, the UO ordering is {element, side, internal side, interface, domain}. There was a comment in the code:
```
// There is one instance in rattlesnake where an elemental user object's finalize depends
// on a side user object having been finalized first :-(
```
I assumed that this was the reason.

Again, I'm not asking for this change in this PR, and like you said, there's a decent chance that other apps are already relying on this ordering, so it's fine to keep. It was just a minor confusion.

Reply:
Oh, that is a very old comment. We possibly should make sure all these mesh related user object types are independent by issuing errors if such a UO tries to get another UO. Users can probably use DomainUserObject or GeneralUserObject to resolve the dependency. If we want some implicit ordering across these mesh UOs, ElementUserObject getting executed last seems a good choice to me.

Reply:
> issuing errors if such a UO tries to get another UO

I bet some apps will get caught by this, but I think there should be ways to avoid hitting this.

Anyway, I added https://github.com/idaholab/moose/issues/23615 for if we ever want to explore this, and I'll resolve this conversation.